climada_petals/entity/exposures/osm_dataloader.py

```diff
@@ -0,0 +1,346 @@
+
+        if isinstance(area,shapely.geometry.Polygon):
+            lon, lat = area.exterior.coords.xy
+            lat_lon_str = " ".join([str(y)+" "+str(x) for y, x in zip(lat, lon)])
+            return f'(poly:"{lat_lon_str}")'
+    
+        return None
+
+    def _insistent_osm_api_query(self, query_clause, read_chunk_size=100000,
+                                 end_of_patience=127):
+        """Runs a single Overpass API query through overpy.Overpass.query.
+        In case of failure it tries again after an ever increasing waiting period.
+        If the waiting period surpasses a given limit an exception is raised.
+
+        Parameters:
+            query_clause (str): the query
+            read_chunk_size (int): paramter passed over to overpy.Overpass.query
+            end_of_patience (int): upper limit for the next waiting period to proceed.
+
+        Returns:
+            result as returned by overpy.Overpass.query
+        """
+        api = overpy.Overpass(read_chunk_size=read_chunk_size)
+        waiting_period = 1
+        while True:
+            try:
+                return api.query(query_clause)
+            except overpy.exception.OverpassTooManyRequests:
+                if waiting_period < end_of_patience:
+                    LOGGER.warning("""Too many Overpass API requests -
+                                   trying again in {waiting_period} seconds """)
+                else:
+                    raise Exception("Overpass API is consistently unavailable")
+            except Exception as exc:
+                if waiting_period < end_of_patience:
+                    LOGGER.warning(f"""{exc}
+                                   Trying again in {waiting_period} seconds""")
+                else:
+                    raise Exception("The Overpass API is consistently unavailable")
+            time.sleep(waiting_period)
+            waiting_period *= 2
+
+    def _overpass_query_string(self):
+        return f'[out:json][timeout:180];(nwr{self.condition}{self.area};(._;>;););out;'
+
+    def _assemble_from_relations(self, result):
+        """
+        pick out those nodes and ways from result instance that belong to
+        relations. Assemble relations into gdfs. Keep track of which nodes
+        and ways have been "used up" already
+
+        Parameters
+        ---------
+        result : overpy.Overpass result object
+
+        Returns
+        -------
+        nodes_taken : list
+            node-ids that have been used to construct relations.
+            Not "available" anymore for further constructions
+        ways_taken : list
+            way-ids that have been used to construct relations.
+            Not "available" anymore for further constructions
+        gdf_rels : gpd.GeoDataFrame
+            gdf with relations that were assembled from result object
+        """
+
+        nodes_taken = []
+        ways_taken = []
+        data_geom = []
+        data_id = []
+        data_tags = []
```

Comment:
Having multiple lists and appending to them is dangerous, because you might end up with lists of different lengths. Appending is also quite inefficient when objects or lists get large. Could you maybe make the loop that appends on these lists a separate function, so the lists can be assigned to the results immediately through a list comprehension?